sklearn/preprocessing/_discretization.py

```diff
@@ -274,54 +240,32 @@ def transform(self, X):
             Data in the binned space.
         """
         check_is_fitted(self, ["bin_edges_"])
-        X = self._validate_X_post_fit(X)
-
-        Xt = _transform_selected(X, self._transform, self.dtype,
-                                 self.transformed_features_, copy=True,
-                                 retain_order=True)
-
-        if self.encode == 'ordinal':
-            return Xt
-
-        # Only one-hot encode discretized features
-        mask = np.ones(X.shape[1], dtype=bool)
-        if self.ignored_features is not None:
-            mask[self.ignored_features] = False
-
-        encode_sparse = self.encode == 'onehot'
-        return OneHotEncoder(n_values=self.n_bins_[mask],
-                             categorical_features='all'
-                             if self.ignored_features is None else mask,
-                             sparse=encode_sparse).fit_transform(Xt)
-
-    def _validate_X_post_fit(self, X):
-        X = check_array(X, dtype='numeric')
 
+        X = check_array(X, dtype=FLOAT_DTYPES)
         n_features = self.n_bins_.shape[0]
         if X.shape[1] != n_features:
             raise ValueError("Incorrect number of features. Expecting {}, "
                              "received {}.".format(n_features, X.shape[1]))
-        return X
-
-    def _transform(self, X):
-        """Performs transformation on X, with no ignored features."""
-        trans = self.transformed_features_
 
-        bin_edges = self.bin_edges_[trans]
-        for jj in range(X.shape[1]):
+        Xt = X.copy()
```

Comment:
Why no using the copy parameter of `check_array`?
It would avoid a double copy if `X` is not a float array.

This applies also for `inverse_transform`.